### PR TITLE
Add Elasticsearch 5.4.0 Cluster

### PR DESCRIPTION
**Pull request by daddz**
Contains:
- scalable master nodes
- scalable data nodes
- scalable client nodes

### CODE REVIEW

**Comment by rawmind0** [comment]
Hi @daddz , thanks for the PR and to contribute.....

Nice package, what do you think to remove kibana from here and to use the already created one??

**Comment by daddz** [comment]
Sure I can remove it if you prefer it that way. 
The idea was to build a simple all-in-one package without external config images.

**Comment by rawmind0** [comment]
I could understand your point @daddz , but i think it could be better to maintain every software by its way. Also it could be useful just elastisearch, due to not everybody that use it want to use kibana,..., or yes but up to them. ;P 

Kibana 5.3.0 update was already merged yesterday with this PR, https://github.com/rancher/community-catalog/pull/484

**Comment by daddz** [comment]
@rawmind0 I removed kibana! :)

**Comment by daddz** [comment]
updated to latest release

@rawmind0 is this ready to merge now?

**Comment by daddz** [comment]
updated to 5.4.0

**Comment by rawmind0** [comment]
Hi @daddz ...

I'm trying to deploy your package and i'm getting vm.max_map_count error...
```
2017-5-11 01:35:13[2017-05-10T23:35:13,708][INFO ][o.e.d.DiscoveryModule    ] [es-cluster-es-master-2] using discovery type [zen]
2017-5-11 01:35:13[2017-05-10T23:35:13,792][INFO ][o.e.x.m.j.p.l.CppLogMessageHandler] [controller/58] [Main.cc@128] controller (64 bit): Version 5.4.0 (Build 120b96fa7f6fa7) Copyright (c) 2017 Elasticsearch BV
2017-5-11 01:35:15[2017-05-10T23:35:15,204][INFO ][o.e.n.Node               ] [es-cluster-es-master-2] initialized
2017-5-11 01:35:15[2017-05-10T23:35:15,207][INFO ][o.e.n.Node               ] [es-cluster-es-master-2] starting ...
2017-5-11 01:35:15[2017-05-10T23:35:15,859][INFO ][o.e.t.TransportService   ] [es-cluster-es-master-2] publish_address {10.42.248.145:9300}, bound_addresses {[::]:9300}
2017-5-11 01:35:15[2017-05-10T23:35:15,874][INFO ][o.e.b.BootstrapChecks    ] [es-cluster-es-master-2] bound or publishing to a non-loopback or non-link-local address, enforcing bootstrap checks
2017-5-11 01:35:15ERROR: bootstrap checks failed
2017-5-11 01:35:15max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]
2017-5-11 01:35:15[2017-05-10T23:35:15,911][INFO ][o.e.n.Node               ] [es-cluster-es-master-2] stopping ...
2017-5-11 01:35:15[2017-05-10T23:35:15,941][INFO ][o.e.x.m.j.p.NativeController] Native controller process has stopped - no new native processes can be started
2017-5-11 01:35:16[2017-05-10T23:35:16,158][INFO ][o.e.n.Node               ] [es-cluster-es-master-2] stopped
2017-5-11 01:35:16[2017-05-10T23:35:16,159][INFO ][o.e.n.Node               ] [es-cluster-es-master-2] closing ...
2017-5-11 01:35:16[2017-05-10T23:35:16,221][INFO ][o.e.n.Node               ] [es-cluster-es-master-2] closed
```

It seems that it's mandatory to increase vm.max_map_count to 262144 at host  level... [elastic-issue](https://github.com/docker-library/elasticsearch/issues/111)

I think you need another privileged sidekick just to set ```sudo sysctl -w vm.max_map_count=262144```

Otherwise, services are not running distributed between hosts... You need another label to assure that... See Change requests... :)

File: templates/es-cluster/0/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,103 @@
+version: '2'
+services:
+    es-master:
+        labels:
```
Comment (rawmind0) [inline templates/es-cluster/0/docker-compose.yml]: Set label to distribute the services, and correct format
```
++            io.rancher.scheduler.affinity:container_label_soft_ne: io.rancher.stack_service.name=$${stack_name}/$${service_name}
++            io.rancher.container.hostname_override: container_name
++            io.rancher.sidekicks: es-storage
```
Delete this format
```
--            - io.rancher.container.hostname_override=container_name
--            - io.rancher.sidekicks=es-storage
```

File: templates/es-cluster/0/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                hard: 65536
+        mem_limit: ${master_mem_limit}
+        mem_swappiness: 0
+        cap_add: 
+            - IPC_LOCK
+        volumes_from: 
+            - es-storage
+
+    es-data:
+        labels:
```
Comment (rawmind0) [inline templates/es-cluster/0/docker-compose.yml]: Set label to distribute the services, and correct format
```
++            io.rancher.scheduler.affinity:container_label_soft_ne: io.rancher.stack_service.name=$${stack_name}/$${service_name}
++            io.rancher.container.hostname_override: container_name
++            io.rancher.sidekicks: es-storage
```
Delete this format
```
--            - io.rancher.container.hostname_override=container_name
--            - io.rancher.sidekicks=es-storage
```

File: templates/es-cluster/0/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        mem_swappiness: 0
+        cap_add: 
+            - IPC_LOCK
+        volumes_from: 
+            - es-storage
+        depends_on: 
+            - es-master
+
+    es-client:
+        labels:
```
Comment (rawmind0) [inline templates/es-cluster/0/docker-compose.yml]: Set label to distribute the services, and correct format
```
++            io.rancher.scheduler.affinity:container_label_soft_ne: io.rancher.stack_service.name=$${stack_name}/$${service_name}
++            io.rancher.container.hostname_override: container_name
++            io.rancher.sidekicks: es-storage
```
Delete this format
```
--            - io.rancher.container.hostname_override=container_name
--            - io.rancher.sidekicks=es-storage
```

**Comment by daddz** [comment]
I added the label fixes.

Regarding the max_map_count: 
Do you think it's a good idea to automatically mess with host settings through a deployment? It might affect other software a user is running on the host.

Wouldn't it be better to make the user aware that he has to set this for all desired hosts e.g. in the readme or somewhere else?

**Comment by daddz** [comment]
Added a note to the readme for max_map_count

**Comment by rawmind0** [comment]
Hi @daddz ... Thanks for your update....

I understand your concerns about modify vm.max_map_count parameter, but in this case i think it doesn't have side effects on your hosts. 

Add an advice in README may work, but i think that it will generate few issues from users....

I've prepared a sidekick just to check and set sysctl param. I've tested it and it's working. 
What do you think if we change docker-compose.yml to docker-compose.yml.tpl and we set a question about update or not sysctl param?? 

Take a look at [my catalog](https://github.com/rawmind0/service-catalog/tree/master/templates/es-cluster/0) updated version.... :)

**Comment by daddz** [comment]
@rawmind0: that's even better! Thanks!

I integrated your changes

**Comment by daddz** [comment]
One last question: is the storage-volume safe when upgrading?
Let's say you deploy 5.4.0 now and want to upgrade to 5.4.1 later with the updated catalog.
If we update the version of the es-storage container will the volume's stick around or will new ones be created and therefore the data is lost?
I'm not sure how this behaves with the per_container setting when the image version gets changed.

File: templates/es-cluster/0/docker-compose.yml.tpl [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    es-storage:
+        labels:
+            io.rancher.container.start_once: true
+        image: docker.elastic.co/elasticsearch/elasticsearch:5.4.0
+        entrypoint: /bin/true
+        volumes: 
+            - es-storage-volume:/usr/share/elasticsearch/data
+
+    {{- if eq .Values.UPDATE_SYSCTL "true" }}
+    es-sysctl:
```
Comment (rawmind0) [inline templates/es-cluster/0/docker-compose.yml.tpl]: Disable network in sidekick
```
+        network_mode: none
```

File: templates/es-cluster/0/docker-compose.yml.tpl [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        mem_limit: ${client_mem_limit}
+        mem_swappiness: 0
+        cap_add: 
+            - IPC_LOCK
+        volumes_from: 
+            - es-storage
+        depends_on: 
+            - es-master
+
+    es-storage:
```
Comment (rawmind0) [inline templates/es-cluster/0/docker-compose.yml.tpl]: Disable network in sidekick
```
+        network_mode: none
```

**Comment by rawmind0** [comment]
For es-storage, i would use a generic and tiny image like busybox or rawmind/alpine-volume ... 
When you upgrade es version, es-storage version shouldn't change to maintain volumes...

Something like that....
```
    es-storage:
      network_mode: none
      labels:
        io.rancher.container.start_once: true
      image: busybox
      entrypoint: /bin/true
      volumes: 
        - es-storage-volume:/usr/share/elasticsearch/data
 volumes:
   es-storage-volume:
     driver: local
     per_container: true
```
or that..
```
    es-storage:
      network_mode: none
      labels:
        io.rancher.container.start_once: true
      image: rawmind/alpine-volume:0.0.2-1
      environment:
        - SERVICE_UID=0
        - SERVICE_GID=0
        - SERVICE_VOLUME=/usr/share/elasticsearch/data
      volumes: 
        - es-storage-volume:/usr/share/elasticsearch/data
 volumes:
   es-storage-volume:
     driver: local
     per_container: true
```

BTW, sorry i miss to request you before to add ```network_mode: none``` to sidekicks...requested...

**Comment by daddz** [comment]
@rawmind0: I used your volume image as it allowed to set uid/gid.
The ES images use the elasticsearch user:
```
bash-4.3$ id elasticsearch
uid=1000(elasticsearch) gid=1000(elasticsearch) groups=1000(elasticsearch)
```
so I used those values.

**Comment by rawmind0** [comment]
I was checking tab and spaces...Thanks to fix it..

Is your es service launched by a user with uid and gid = 1000?? Sure that they are not 0??
```
+            - SERVICE_UID=1000
+            - SERVICE_GID=1000
```

**Comment by daddz** [comment]
The images provided for elasticsearch don't run as root and create a "elasticsearch" user that happens to be the first and so is uid=1000 gid=1000

**Comment by rawmind0** [comment]
Sorry...I missed your comment about it....Pretty cool!!

Tested and working ok....LGTM..

Thanks for contribution @daddz... :)

**Comment by rawmind0** [comment]
Duplicate https://github.com/rancher/community-catalog/pull/503

**Pull request merged**